src/ServiceControl.Monitoring.Tests/PerformanceTests.cs

```diff
@@ -0,0 +1,245 @@
+    using Timings;
+
+    [Category("Performance")]
+    public class PerformanceTests
+    {
+        EndpointRegistry endpointRegistry;
+        CriticalTimeStore criticalTimeStore;
+        ProcessingTimeStore processingTimeStore;
+        RetriesStore retriesStore;
+        QueueLengthStore queueLengthStore;
+        Func<Task> GetMonitoredEndpoints;
+
+        [SetUp]
+        public void Setup()
+        {
+            endpointRegistry = new EndpointRegistry();
+            criticalTimeStore = new CriticalTimeStore();
+            processingTimeStore = new ProcessingTimeStore();
+            retriesStore = new RetriesStore();
+            queueLengthStore = new QueueLengthStore(new QueueLengthCalculator());
+
+            var monitoredEndpointsModule = new MonitoredEndpointsModule(endpointRegistry, criticalTimeStore, processingTimeStore, retriesStore, queueLengthStore)
+            {
+                Context = new NancyContext()
+            };
+
+            var dictionary = monitoredEndpointsModule.Routes.ToDictionary(r => r.Description.Path, r => r.Action);
+
+            GetMonitoredEndpoints = () => dictionary["/monitored-endpoints"](new object(), new CancellationToken(false));
+        }
+
+        [TestCase(10, 5, 4)]
+        [TestCase(1000, 5, 4)]
+        [TestCase(1000, 100, 4)]
+        public async Task TestQueueLengthStore(int numberOfEndpoints, int numberOfReceivedSessionsPerEndpoint, int numberOfSentSessionsPerEndpoint)
+        {
+            const int totalQueryCount = 200;
+            var sessionCount = numberOfEndpoints * numberOfSentSessionsPerEndpoint;
+            var allSessions = Enumerable.Range(0, sessionCount).Select(i => $"Session-{i}").ToArray();
+            var allEndpoints = Enumerable.Range(0, numberOfEndpoints).Select(i => $"Endpoint-{i}").ToArray();
+
+            var endpointReporters = new Task<LongHistogram>[allEndpoints.Length];
+
+            var source = new CancellationTokenSource();
+
+            var random = new Random();
+            for (var i = 0; i < allEndpoints.Length; i++)
+            {
+                var sentSessions = Enumerable.Range(i * numberOfSentSessionsPerEndpoint, numberOfSentSessionsPerEndpoint)
+                    .Select(index => allSessions[index])
+                    .ToArray();
+
+                var receivedSessions = Enumerable.Range(0, numberOfReceivedSessionsPerEndpoint)
+                    .Select(_ => random.Next(sessionCount))
+                    .Select(si => allSessions[si]).ToArray();
+
+                endpointReporters[i] = BuildQueueLengthReporter(allEndpoints[i], receivedSessions, sentSessions, queueLengthStore, source);
+            }
+
+            var snapshoter = Task.Run(async () =>
+            {
+                for (var i = 0; i < totalQueryCount; i++)
+                {
+                    queueLengthStore.SnapshotCurrentQueueLengthEstimations(DateTime.UtcNow);
+
+                    await Task.Delay(TimeSpan.FromMilliseconds(10));
```

Comment:
Is this 10s the same as the below 10s? Should this be a variable/const somewhere?